# rarepedsim

Generation of sequence-based data for pedigree-segregating Mendelian or Complex traits

* Published on Bioinformatics Journal, [link](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4757949/)

## Implementation, Installation and User Guide
* Refer to [Documentation](doc/doc_RarePedSim.pdf)
	* Download [source code](src/) and [example data](data/) 
* Project also hosted on the [website](http://bioinformatics.org/simped/rare/)

## Introduction
RarePedSim (Rare-variant Pedigree-based Simulator), is a program to simulate region/gene-level genotype and phenotype data for complex and Mendelian traits for any given pedigree structure. Using a genetic model, sequence variant data can be generated either conditionally or unconditionally on pedigree membersâ€™ qualitative or quantitative phenotypes. Additionally, qualitative or quantitative traits can be generated conditional on variant data. Sequence data can either be simulated using realistic population demographic models or obtained from sequence-based studies. Variant sites can be annotated with positions, allele frequencies and functionality. For rare variants, RarePedSim is the only program that can efficiently generate both genotypes and phenotypes, regardless of pedigree structure. Data generated by RarePedSim are in standard Linkage file (.ped) and Variant Call (.vcf) formats, ready to be used for a variety of purposes, including evaluation of type I error and power, for association methods including mixed models and linkage analysis methods.